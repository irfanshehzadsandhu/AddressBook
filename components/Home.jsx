import $ from "jquery";
import React from "react";
import { connect } from "react-redux";
import { fetchUsers, fetchingUsersBegin } from "../actions/usersAction";
import User from "./User.jsx";
import Modal from "./Modal.jsx";

class App extends React.Component {
  constructor() {
    super();
    this.page = 1;
    this.offset = 30;
    this.handleScroll = this.handleScroll.bind(this);
  }

  getPage() {
    return this.page;
  }
  getOffset() {
    return this.offset;
  }

  setPage() {
    return (this.page = this.page + 1);
  }

  openModel(e) {
    let parentNode = e.target.parentNode.parentNode;
    $("#street-number").html(parentNode.getAttribute("data-street-number"));
    $("#street-name").html(parentNode.getAttribute("data-street-name"));
    $("#city").html(parentNode.getAttribute("data-city"));
    $("#state").html(parentNode.getAttribute("data-state"));
    $("#country").html(parentNode.getAttribute("data-country"));
    $("#postcode").html(parentNode.getAttribute("data-postcode"));
    $("#userDetailsModal").modal("show");
  }

  componentDidMount() {
    this.props.dispatch(fetchingUsersBegin());
    this.props.dispatch(
      fetchUsers(this.getPage(), this.getOffset(), this.props.nationality)
    );
    window.addEventListener("scroll", this.handleScroll);
  }

  componentWillUnmount() {
    window.removeEventListener("scroll", this.handleScroll);
  }

  handleScroll() {
    const windowHeight =
      "innerHeight" in window
        ? window.innerHeight
        : document.documentElement.offsetHeight;
    const body = document.body;
    const html = document.documentElement;
    const docHeight = Math.max(
      body.scrollHeight,
      body.offsetHeight,
      html.clientHeight,
      html.scrollHeight,
      html.offsetHeight
    );
    const windowBottom = windowHeight + window.pageYOffset;
    if (windowBottom >= docHeight) {
      const { loading, dispatch } = this.props;
      if (loading == false) {
        dispatch(
          fetchUsers(this.setPage(), this.getOffset(), this.props.nationality)
        );
      }
    } else {
      console.log("Reaching bottom");
    }
  }

  render() {
    const { error, loading, usersList } = this.props;

    if (error) {
      return <div>Error! {error.message}</div>;
    }

    if (loading) {
      return (
        <div style={{ marginTop: 300, marginLeft: 700 }}>
          <img src="../assets/images/tenor.gif" />
        </div>
      );
    }
    return (
      <div>
        <Modal />
        <table className="table">
          <thead>
            <tr>
              <th></th>
              <th>First Name</th>
              <th>Last Name</th>
              <th>Email</th>
              <th>View</th>
            </tr>
          </thead>
          <tbody>
            {usersList.map((user, index) => {
              return (
                <User key={index} user={user} openModel={this.openModel} />
              );
            })}
          </tbody>
        </table>
      </div>
    );
  }
}
function mapStateToProps(state) {
  return {
    usersList: state.users.usersList,
    loading: state.users.loading,
    error: state.users.error,
    nationality: state.users.nationality,
  };
}
export default connect(mapStateToProps)(App);
