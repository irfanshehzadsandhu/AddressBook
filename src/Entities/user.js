import Location from "./location.js";
export default class User {
  constructor(
    id,
    userName,
    thumbnail,
    image,
    firstName,
    lastName,
    email,
    location,
    phone,
    cell
  ) {
    this.id = id;
    this.userName = userName;
    this.image = image;
    this.thumbnail = thumbnail;
    this.firstName = firstName;
    this.lastName = lastName;
    this.email = email;
    this.location = location;
    this.phone = phone;
    this.cell = cell;
  }

  fullName() {
    return this.firstName + " " + this.lastName;
  }

  static createFromObj(obj) {
    return new User(
      obj["login"]["uuid"],
      obj["login"]["username"],
      obj["picture"]["thumbnail"],
      obj["picture"]["large"],
      obj["name"]["first"],
      obj["name"]["last"],
      obj["email"],
      Location.createLocFromObj(obj["location"]),
      obj["phone"],
      obj["cell"]
    );
  }
}
