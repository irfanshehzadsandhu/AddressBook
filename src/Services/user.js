import usersLibrary from "../Config";
import PaginatedConfig from "../Utils/paginationConfig.js";
import PaginatedData from "../Utils/paginatedData.js";
import User from "../Entities/user.js";
import { TOTAL_RECORDS } from "../Constants";
export class UserService {
  static async fetchUsersFromEndPoint(page, offset, nationality) {
    const fields = "id,name,email,login,picture,phone,cell,location";
    const res = await fetch(
      usersLibrary.libraryEndPoint +
        "?page=" +
        page +
        "&results=" +
        offset +
        "&nat=" +
        nationality +
        "&inc=" +
        fields
    );
    const json = await res.json();
    const paginatedData = new PaginatedData(
      new PaginatedConfig(page, offset),
      TOTAL_RECORDS
    );
    json.results.map((result) => {
      paginatedData.addItem(User.createFromObj(result));
    });
    return paginatedData.paginatedItems();
  }
}
