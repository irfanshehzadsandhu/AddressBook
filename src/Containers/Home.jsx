import React from "react";
import { connect } from "react-redux";
import "../../assets/stylesheets/Home.css";
import { Row, Col } from "antd";
import { fetchUsers, fetchingUsersBegin } from "../Redux/Actions/usersAction";
import User from "../Components/User.jsx";
import UserModal from "../Components/Modal.jsx";

class Home extends React.Component {
  getUsers() {
    const { paginationInfo, nationality } = this.props;
    if (paginationInfo.hasNextPage) {
      this.props.fetchUsers(
        paginationInfo.nextPage,
        paginationInfo.perPage,
        nationality
      );
    }
  }
  componentDidMount() {
    this.props.fetchingUsersBegin();
    this.getUsers();
    this.handleScrollToBottom();
  }

  componentWillUnmount() {
    window.removeEventListener("scroll", this.handleScroll);
  }

  handleScrollToBottom() {
    const { paginationInfo, loading } = this.props;
    window.onscroll = async () => {
      if (loading || !paginationInfo.hasNextPage) return;
      if (
        window.innerHeight + document.documentElement.scrollTop ===
        document.documentElement.scrollHeight
      ) {
        console.log("loading more data");
        this.getUsers();
      }
    };
  }

  render() {
    const { error, loading, usersList, paginationInfo } = this.props;

    if (error) {
      return <div>Error! {error.message}</div>;
    }

    if (loading) {
      return (
        <div style={{ marginTop: 300, marginLeft: 700 }}>
          <img src="../assets/images/tenor.gif" />
        </div>
      );
    }
    return (
      <>
        <Row>
          <Col span={6}>Avatar</Col>
          <Col span={6}>Name</Col>
          <Col span={6}>Email</Col>
          <Col span={6}>Details</Col>
        </Row>
        {usersList.map((user, index) => {
          return <User key={index} user={user} />;
        })}
        {paginationInfo.hasNextPage == false ? <div>No more users</div> : ""}
      </>
    );
  }
}
const mapDispatchToProps = {
  fetchUsers,
  fetchingUsersBegin,
};
function mapStateToProps(state) {
  return {
    usersList: state.users.usersList,
    loading: state.users.loading,
    error: state.users.error,
    nationality: state.users.nationality,
    paginationInfo: state.users.paginationInfo,
  };
}
export default connect(mapStateToProps, mapDispatchToProps)(Home);
