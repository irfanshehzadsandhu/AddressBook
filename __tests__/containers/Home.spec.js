import React from "react";
import { Provider } from "react-redux";
import thunk from "redux-thunk";
import { mount } from "enzyme";
import configureStore from "redux-mock-store";
import sinon from "sinon";
import Home from "../../src/Containers/Home.jsx";
import { UserService } from "../../src/Services/user.js";
import UserFactory from "../../tests/factories/user.js";

const mockStore = configureStore([thunk]);

describe("My Connected React-Redux Component", () => {
  let store;
  let component;
  beforeEach(() => {
    Object.defineProperty(window, "matchMedia", {
      value: () => {
        return {
          matches: false,
          addListener: () => {},
          removeListener: () => {},
        };
      },
    });
    store = mockStore({
      users: {
        usersList: [],
        loading: false,
        error: null,
        nationality: "us",
        paginationInfo: {
          currentPage: 0,
          nextPage: 1,
          hasNextPage: true,
          hasPrevPage: false,
          perPage: 30,
        },
        selectedUser: null,
        cachedUsersList: [],
      },
    });
    sinon.stub(UserService, "fetchUsersFromEndPoint").returns({
      data: [UserFactory.createUser(), UserFactory.createUser()],
      paginationInfo: {
        currentPage: 1,
        nextPage: 2,
        hasNextPage: true,
        hasPrevPage: false,
        perPage: 30,
      },
    });
    component = mount(
      <Provider store={store}>
        <Home />
      </Provider>
    );
  });

  it("should render with given state from Redux store", () => {
    expect(component).toMatchSnapshot();
  });
});
