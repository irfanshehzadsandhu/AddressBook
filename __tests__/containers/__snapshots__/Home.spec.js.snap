// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`My Connected React-Redux Component should render with given state from Redux store 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(Home)>
    <Home
      error={null}
      fetchUsers={[Function]}
      fetchingUsersBegin={[Function]}
      loading={false}
      nationality="us"
      paginationInfo={
        Object {
          "currentPage": 0,
          "hasNextPage": true,
          "hasPrevPage": false,
          "nextPage": 1,
          "perPage": 30,
        }
      }
      selectUser={[Function]}
      selectedUser={null}
      unSelectUser={[Function]}
      usersList={Array []}
    />
  </Connect(Home)>
</Provider>
`;
